---
title: Python post
author: Jakson Johnson
date: '2019-12-15'
slug: python-post
categories: []
tags: []
description: ''
---



<p>Python is one of the most popular programming languages in the world, and it is getting progressively more popular every day! Python is great for lots of things, but in this example I will show how it can be used in the realm of biology for the purpose of aligning sequences of DNA. The Needlman-Wunsch algorithim was developed in 1970 for the purpose of finding the optimal alignment of two different DNA sequences. In a nut shell, it generates an alignment matrix by assigning numerical score values for nucelotide matches, mismatches, and gaps. The purpose of aligning DNA sequences is to identify things like short tandem repeats, which are used in crime labs to identify suspects, or to simply determine how well conserved certain sections of a genome are amongst different species (very important for generating phylogenies!). Here we have written some code in python that creates an alignment matrix for two, simple nucleotide sequences using the needleman-wunsch algorithim.</p>
<pre class="r"><code>library(reticulate)</code></pre>
<pre><code>## Warning: package &#39;reticulate&#39; was built under R version 3.5.3</code></pre>
<pre class="python"><code>import numpy as np

def match_fun(a,b):
    if a==b:
        score=1 #match score
    else:
        score=-1 #mismatch score
    return score # Make a score matrix with these two sequences
    
seq1 = &quot;ATGAT&quot;
seq2 = &quot;AAGGAT&quot;


def needleman(seq1,seq2,gap_penalty=-1):  # Needleman-Wunsch algo in a function
    m=len(seq1) #length of horizontal sequence
    n=len(seq2) #length of vertical sequence
    score=np.empty(shape=[n+1,m+1]) #array to hold scores
   
    for j in range(0, m + 1): score[0][j] = gap_penalty * j
    for i in range(0, n + 1): score[i][0] = gap_penalty * i
    
    for i in range(1, n + 1): 
        for j in range(1, m + 1):
            insert = score[i - 1][j] + gap_penalty 
            delete = score[i][j - 1] + gap_penalty 
            match = score[i - 1][j - 1] + match_fun(seq1[j-1], seq2[i-1])
            score[i][j] = max(match, delete, insert)
    return score

needleman(seq1,seq2)</code></pre>
<pre><code>## array([[ 0., -1., -2., -3., -4., -5.],
##        [-1.,  1.,  0., -1., -2., -3.],
##        [-2.,  0.,  0., -1.,  0., -1.],
##        [-3., -1., -1.,  1.,  0., -1.],
##        [-4., -2., -2.,  0.,  0., -1.],
##        [-5., -3., -3., -1.,  1.,  0.],
##        [-6., -4., -2., -2.,  0.,  2.]])</code></pre>
<p>As you can see, a needleman-wunsch matrix has been generated for the two sample nucleotide sequences we defined (“ATGAT” and “AAGGAT”). Now I could tell you how to interpret this matrix, but why bore you with that when we can just write some more code that will interpret the matrix for us?</p>
<pre class="python"><code>mat=needleman(seq1,seq2)

def get_alignment(seq1,seq2,mat,gap_penalty=-1):
    j=len(seq1) #length of horizontal sequence
    i=len(seq2) #length of vertical sequence
    AlignA=&quot;&quot;   #empty strings to hold alignments
    AlignB=&quot;&quot;
    while i &gt; 0 or j &gt; 0:  #start at the bottom right corner: if from diagonal, align sequences
        if i &gt; 0 and j &gt; 0 and mat[i,j]==mat[i-1][j-1]+match_fun(seq1[j-1],seq2[i-1]):
            AlignA = seq1[j-1] + AlignA
            AlignB = seq2[i-1] + AlignB
            i -= 1
            j -= 1
        elif j &gt; 0 and mat[i,j]==mat[i][j-1]+gap_penalty: #if from above, put gap in vertical sequence
            AlignA = seq1[j-1] + AlignA
            AlignB = &quot;-&quot; + AlignB
            j -= 1
        else:                             #if from the left, put gap in horizontal sequence
            AlignA = &quot;-&quot; + AlignA
            AlignB = seq2[i-1]+AlignB
            i -= 1
    return AlignA, AlignB #return both alignments
    
get_alignment(seq1,seq2,mat)</code></pre>
<pre><code>## (&#39;-ATGAT&#39;, &#39;AAGGAT&#39;)</code></pre>
<p>As you can see here, this function generates the optimal alignment for the two sequences as defined by the needleman-wunsch matrix. Hope this goes to show that Python is useful to more people than just those with a computer science degree!</p>
